service/evaluation/evaluation.go

```diff
@@ -0,0 +1,845 @@
+
+	}
+}
+
+// evaluateSubcontrol evaluates the sub-controls, e.g., OPS-13.2
+func (s *Service) evaluateSubcontrol(toe *orchestrator.TargetOfEvaluation, categoryName, controlId, parentSchedulerTag string) {
+	var (
+		eval              *evaluation.EvaluationResult
+		err               error
+		assessmentResults []*assessment.AssessmentResult
+	)
+
+	// Get metrics from control and sub-controls
+	metrics, err := s.getAllMetricsFromControl(toe.GetCatalogId(), categoryName, controlId)
+	if err != nil {
+		log.Errorf("could not get metrics for controlID '%s' and Cloud Service '%s' from Orchestrator: %v", controlId, toe.GetCloudServiceId(), err)
+		// If the parentSchedulerTag is not empty, we have do decrement the WaitGroup so that the parent control can also be evaluated when all sub-controls are evaluated.
+		if parentSchedulerTag != "" && s.wg[parentSchedulerTag] != nil {
+			s.wg[parentSchedulerTag].wgMutex.Lock()
+			s.wg[parentSchedulerTag].wg.Done()
+			s.wg[parentSchedulerTag].wgMutex.Unlock()
+		}
+		return
+	}
+
+	if len(metrics) != 0 {
+		// Get latest assessment_results by resource_id filtered by
+		// * cloud service id
+		// * metric ids
+		assessmentResults, err = api.ListAllPaginated(&orchestrator.ListAssessmentResultsRequest{
+			FilteredCloudServiceId: &toe.CloudServiceId,
+			FilteredMetricId:       getMetricIds(metrics),
+			LatestByResourceId:     util.Ref(true),
+		}, s.orchestratorClient.ListAssessmentResults, func(res *orchestrator.ListAssessmentResultsResponse) []*assessment.AssessmentResult {
+			return res.Results
+		})
+
+		if err != nil {
+			// We let the scheduler running if we do not get the assessment results from the orchestrator, maybe it is only a temporary network problem
+			log.Errorf("could not get assessment results for Cloud Service ID '%s' and MetricIds '%s' from Orchestrator: %v", toe.GetCloudServiceId(), getMetricIds(metrics), err)
+		} else if len(assessmentResults) == 0 {
+			// We let the scheduler running if we do not get the assessment results from the orchestrator, maybe it is only a temporary network problem
+			log.Debugf("no assessment results for Cloud Service ID '%s' and MetricIds '%s' available", toe.GetCloudServiceId(), getMetricIds(metrics))
+		}
+	} else {
+		log.Debugf("no metrics are available for the given control")
+	}
+
+	// Here the actual evaluation takes place. For every resource_id we check if the asssessment results are compliant. If the latest assessment result per resource_id is not compliant the whole evaluation status is set to NOT_COMPLIANT. Furthermore, all non-compliant assessment_result_ids are stored in a separate list.
+
+	// Get a map of the assessment results, so that we have all assessment results for a specific resource_id and metric_id together for evaluation
+	assessmentResultsMap := getAssessmentResultMap(assessmentResults)
+	for _, result := range assessmentResultsMap {
```

Comment:
better name would probably be `results`, since its a list, otherwise its confusing.